File: src/SettingsPage.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;

namespace OpusScreen
{
    /// <summary>
    /// Base commune a toutes les pages de reglages.
    ///
    /// Chaque page ne s'occupe que de son domaine et empile ses lignes ; le
    /// positionnement, le defilement et les marges sont geres ici une bonne fois.
    /// C'est ce qui permet d'ajouter un reglage en une ligne sans toucher a la mise
    /// en page.
    ///
    /// L'empilement est recalcule a chaque redimensionnement plutot que fige a la
    /// construction : les textes explicatifs ne connaissent leur nombre de lignes
    /// qu'une fois la largeur reelle connue.
    /// </summary>
    public abstract class SettingsPage : Panel
    {
        protected readonly Settings S;
        protected readonly DisplayController Display;
        protected readonly Action Push;
        protected bool Loading;

        private static int Pad { get { return Theme.Px(24); } }

        private class Item
        {
            public Control Control;
            public int Gap;
            public string WrapText;   // non nul pour les textes explicatifs a remesurer
        }

        private readonly List<Item> _items = new List<Item>();
        private int _lastWidth = -1;

        protected SettingsPage(Settings s, DisplayController display, Action push)
        {
            S = s;
            Display = display;
            Push = push;

            BackColor = Theme.Card;
            AutoScroll = true;
        }

        /// <summary>Ajoute un controle a la suite du precedent.</summary>
        protected T Add<T>(T c, int gapBefore) where T : Control
        {
            Item it = new Item();
            it.Control = c;
            it.Gap = gapBefore;
            _items.Add(it);
            Controls.Add(c);
            return c;
        }

        protected T Add<T>(T c) where T : Control { return Add(c, Theme.SpaceSm); }

        protected void Section(string title)
        {
            Add(UiKit.SectionTitle(title), _items.Count == 0 ? Theme.Px(6) : Theme.SpaceXl);
            Add(UiKit.Divider(), Theme.Px(2));
        }

        /// <summary>Texte explicatif, dont la hauteur suit le nombre de lignes reellement occupees.</summary>
        protected Label Note(string text)
        {
            Label l = UiKit.Caption(text);
            Item it = new Item();
            it.Control = l;
            it.Gap = Theme.SpaceXs;
            it.WrapText = text;
            _items.Add(it);
            Controls.Add(l);
            return l;
        }

        protected int ContentWidth
        {
            get { return Math.Max(Theme.Px(160), ClientSize.Width - Pad * 2 - (VerticalScroll.Visible ? Theme.Px(16) : 0)); }
        }

        protected override void OnResize(EventArgs e)
        {
            base.OnResize(e);
            if (ClientSize.Width != _lastWidth) Relayout();
        }

        protected override void OnVisibleChanged(EventArgs e)
        {
            base.OnVisibleChanged(e);
            if (Visible) Relayout();
        }

        /// <summary>Repositionne toute la pile. Peu couteux : quelques dizaines de controles.</summary>
        protected void Relayout()
        {
            int w = ContentWidth;
            if (w <= 160) return;
            _lastWidth = ClientSize.Width;

            SuspendLayout();
            try
            {
                // Dans un panneau qui defile, Top se compte depuis le haut de la zone
                // VISIBLE, pas depuis le haut du contenu. Empiler depuis 14 sans tenir
                // compte du defilement posait donc la pile a partir de l'endroit ou l'on
                // se trouvait : une page defilee de 800 px voyait tout son contenu
                // descendre de 800 px a chaque redisposition - et rafraichir la page
                // Ecrans en declenchait une a chaque reglage.
                int y = Theme.Px(14) + AutoScrollPosition.Y;
                foreach (Item it in _items)
                {
                    if (!it.Control.Visible && it.Control.Height == 0) continue;

                    // Le texte mesure est celui que l'etiquette porte MAINTENANT, et
                    // non celui qu'on lui avait donne a la construction : la page
                    // Avance remplace le sien par l'etat reel du pilote, et sa boite
                    // restait calculee sur la chaine vide qu'elle contenait au depart.
                    if (it.WrapText != null)
                        it.Control.Height = HauteurDuTexte(TexteDe(it.Control, it.WrapText),
                                                           PoliceDe(it.Control), w);
                    else
                        GrandirPourSonTexte(it.Control, w);

                    y += it.Gap;
                    it.Control.SetBounds(Pad, y, w, it.Control.Height);
                    y += it.Control.Height;
                }
            }
            finally { ResumeLayout(true); }
        }

        /// <summary>
        /// Hauteur qu'il faut reserver a un texte pour qu'il tienne en entier.
        ///
        /// Mesuree avec TextRenderer, c'est-a-dire avec le MEME moteur que celui qui
        /// dessinera ces etiquettes. La version precedente mesurait avec
        /// Graphics.MeasureString - GDI+ - alors que Windows Forms peint un Label
        /// avec GDI : les deux ne coupent pas les lignes au meme endroit, et l'ecart
        /// se paie en texte tranche. Sur la page Confort, la boite calculee faisait
        /// 49 pixels pour un texte qui en demandait 60 : deux lignes lisibles, la
        /// troisieme coupee en son milieu, a 100 % comme a 200 %.
        /// </summary>
        private static int HauteurDuTexte(string text, Font police, int width)
        {
            Size besoin = TextRenderer.MeasureText(text, police,
                new Size(Math.Max(60, width), int.MaxValue), TextFormatFlags.WordBreak);
            return besoin.Height + 2;
        }

        /// <summary>
        /// Agrandit une etiquette dont le texte ne tient plus, sans jamais la
        /// retrecir.
        ///
        /// Certaines etiquettes recoivent leur texte APRES la mise en page - l'etat
        /// du pilote Intel, le resume du dernier test, le nombre d'applications au
        /// mixeur. Leur boite avait ete mesuree sur un texte vide : la page Avance en
        /// affichait une haute de deux pixels pour trois lignes d'explication. Le
        /// texte n'etait pas tronque, il etait invisible.
        ///
        /// On ne retrecit pas : une boite dessinee volontairement plus grande - pour
        /// aerer un titre de section - garde sa respiration.
        /// </summary>
        private static void GrandirPourSonTexte(Control c, int width)
        {
            Label l = c as Label;
            if (l == null || l.Text.Length == 0) return;

            int besoin = HauteurDuTexte(l.Text, PoliceDe(l), width);
            if (besoin > l.Height) l.Height = besoin;
        }

        private static Font PoliceDe(Control c)
        {
            return c.Font ?? Theme.Small;
        }

        /// <summary>Le texte affiche, ou celui d'origine si le controle n'en porte pas.</summary>
        private static string TexteDe(Control c, string origine)
        {
            return (c.Text != null && c.Text.Length > 0) ? c.Text : origine;
        }

        /// <summary>Recharge l'affichage a partir des reglages. Appelee a chaque ouverture.</summary>
        public abstract void Sync();

        /// <summary>
        /// Recharge, puis redispose la pile.
        ///
        /// L'ordre compte : Sync pose des textes dont on ne connait la longueur
        /// qu'a cet instant - l'etat du pilote, un resume, un compte. Sans la
        /// redisposition qui suit, ces textes attendaient le prochain
        /// redimensionnement de la fenetre pour obtenir leur place, c'est-a-dire
        /// souvent jamais.
        /// </summary>
        public void SyncAndLayout()
        {
            Sync();
            Relayout();
        }

        /// <summary>Titre affiche en en-tete de la fenetre quand la page est active.</summary>
        public abstract string Title { get; }

        /// <summary>Sous-titre explicatif.</summary>
        public virtual string Subtitle { get { return ""; } }

        protected void Commit()
        {
            if (Loading) return;
            if (Push != null) Push();
            S.Save();
        }

        protected void CommitNoSave()
        {
            if (Loading) return;
            if (Push != null) Push();
        }
    }
}

File: mac/Sources/OpusScreen/UI/SettingsPage.swift
import Foundation
import AppKit

/// Une vue dont l'origine est en haut a gauche.
///
/// AppKit compte depuis le bas ; toute la mise en page de cette application est
/// pensee de haut en bas, comme on lit. Retourner la vue une fois vaut mieux que
/// soustraire des hauteurs a chaque ligne.
public class FlippedView: NSView {
    public override var isFlipped: Bool { true }
}

/// Base commune a toutes les pages de reglages.
///
/// Chaque page ne s'occupe que de son domaine et empile ses lignes ; le
/// positionnement, le defilement et les marges sont geres ici une bonne fois.
/// C'est ce qui permet d'ajouter un reglage en une ligne sans toucher a la mise
/// en page.
///
/// L'empilement est recalcule a chaque redimensionnement plutot que fige a la
/// construction : les textes explicatifs ne connaissent leur nombre de lignes
/// qu'une fois la largeur reelle connue.
public class SettingsPage: NSView {

    public let settings: Settings
    public let display: DisplayController
    public let push: () -> Void
    public var loading = false

    private let scroll = NSScrollView()
    private let stack = FlippedView()

    private struct Item {
        let view: NSView
        let gap: CGFloat
        let wraps: Bool
    }

    private var items: [Item] = []
    private var lastWidth: CGFloat = -1

    private static var pad: CGFloat { Theme.px(24) }

    public init(_ settings: Settings, _ display: DisplayController, _ push: @escaping () -> Void) {
        self.settings = settings
        self.display = display
        self.push = push
        super.init(frame: NSRect(x: 0, y: 0, width: 600, height: 500))

        wantsLayer = true
        layer?.backgroundColor = Theme.card.cgColor

        scroll.hasVerticalScroller = true
        scroll.hasHorizontalScroller = false
        scroll.drawsBackground = false
        scroll.autohidesScrollers = true
        scroll.scrollerStyle = .overlay
        scroll.documentView = stack
        addSubview(scroll)
    }

    required init?(coder: NSCoder) { fatalError() }

    public override var isFlipped: Bool { true }

    // ------------------------------------------------------------------ empilement

    /// Ajoute un controle a la suite du precedent.
    @discardableResult
    public func add<T: NSView>(_ view: T, _ gapBefore: CGFloat? = nil) -> T {
        items.append(Item(view: view, gap: gapBefore ?? Theme.spaceSm, wraps: false))
        stack.addSubview(view)
        return view
    }

    public func section(_ title: String) {
        add(UiKit.sectionTitle(title), items.isEmpty ? Theme.px(6) : Theme.spaceXl)
        add(UiKit.divider(), Theme.px(2))
    }

    /// Texte explicatif, dont la hauteur suit le nombre de lignes reellement occupees.
    @discardableResult
    public func note(_ text: String) -> LabelView {
        let l = UiKit.caption(text)
        items.append(Item(view: l, gap: Theme.spaceXs, wraps: true))
        stack.addSubview(l)
        return l
    }

    public var contentWidth: CGFloat {
        max(Theme.px(160), bounds.width - SettingsPage.pad * 2 - Theme.px(18))
    }

    public override func resizeSubviews(withOldSize oldSize: NSSize) {
        super.resizeSubviews(withOldSize: oldSize)
        scroll.frame = bounds
        if bounds.width != lastWidth { relayout() }
    }

    public override func layout() {
        super.layout()
        scroll.frame = bounds
        if bounds.width != lastWidth { relayout() }
    }

    /// Repositionne toute la pile. Peu couteux : quelques dizaines de controles.
    public func relayout() {
        let w = contentWidth
        guard w > Theme.px(160) else { return }
        lastWidth = bounds.width

        var y = Theme.px(14)
        for item in items {
            if item.view.isHidden { continue }

            // Le texte mesure est celui que l'etiquette porte MAINTENANT, et non
            // celui qu'on lui avait donne a la construction : plusieurs pages
            // remplacent le leur par l'etat reel du systeme, et leur boite
            // resterait calculee sur la chaine vide du depart.
            var height = item.view.frame.height
            if let label = item.view as? LabelView, item.wraps || label.wraps {
                height = label.fittingHeight(for: w)
            } else if let toggle = item.view as? ToggleRow {
                height = toggle.fittingHeight(for: w)
            }

            y += item.gap
            item.view.frame = NSRect(x: SettingsPage.pad, y: y, width: w, height: height)
            item.view.needsLayout = true
            item.view.layoutSubtreeIfNeeded()
            y += height
        }

        y += Theme.px(24)
        stack.frame = NSRect(x: 0, y: 0, width: bounds.width, height: max(y, bounds.height))
    }

    // ------------------------------------------------------------------ contrat

    /// Recharge l'affichage a partir des reglages. Appelee a chaque ouverture.
    public func sync() {}

    /// Recharge, puis redispose la pile.
    ///
    /// L'ordre compte : `sync` pose des textes dont on ne connait la longueur
    /// qu'a cet instant - l'etat du systeme, un resume, un compte. Sans la
    /// redisposition qui suit, ces textes attendaient le prochain
    /// redimensionnement de la fenetre pour obtenir leur place, c'est-a-dire
    /// souvent jamais.
    public func syncAndLayout() {
        sync()
        lastWidth = -1
        relayout()
    }

    /// Titre affiche en en-tete de la fenetre quand la page est active.
    public var pageTitle: String { "" }

    /// Sous-titre explicatif.
    public var pageSubtitle: String { "" }

    public func commit() {
        guard !loading else { return }
        push()
        settings.save()
    }

    public func commitNoSave() {
        guard !loading else { return }
        push()
    }

    /// La fenetre qui contient cette page, pour les boites de dialogue.
    public var hostWindow: NSWindow? { window }
}
